miio/gateway/gateway.py

```diff
@@ -0,0 +1,420 @@
+"""Xiaomi Gateway implementation using Miio protecol."""
+
+import logging
+from enum import IntEnum
+
+import click
+
+from ..click_common import command
+from ..device import Device
+from ..exceptions import DeviceError, DeviceException
+
+_LOGGER = logging.getLogger(__name__)
+
+GATEWAY_MODEL_CHINA = "lumi.gateway.v3"
+GATEWAY_MODEL_EU = "lumi.gateway.mieu01"
+GATEWAY_MODEL_ZIG3 = "lumi.gateway.mgl03"
+GATEWAY_MODEL_AQARA = "lumi.gateway.aqhm01"
+GATEWAY_MODEL_AC_V1 = "lumi.acpartner.v1"
+GATEWAY_MODEL_AC_V2 = "lumi.acpartner.v2"
+GATEWAY_MODEL_AC_V3 = "lumi.acpartner.v3"
+
+
+class GatewayException(DeviceException):
+    """Exception for the Xioami Gateway communication."""
+
+
+class DeviceType(IntEnum):
```

Comment:
This class should probably go into its own file under the subdevices directory. I think the mapping below also belongs there rather than be contained inside the `discover` method below.

Reply:
I have moved them to the `subdevice.py` file